osquery/tables/system/posix/shell_history.cpp

```diff
@@ -12,76 +12,119 @@
 
 #include <osquery/core.h>
 #include <osquery/filesystem/filesystem.h>
-#include <osquery/logger.h>
 #include <osquery/tables.h>
-#include <osquery/tables/system/system_utils.h>
 #include <osquery/tables/system/posix/shell_history.h>
-#include <osquery/utils/conversions/split.h>
+#include <osquery/tables/system/system_utils.h>
 #include <osquery/utils/system/system.h>
 
 namespace osquery {
 namespace tables {
 
 const std::vector<std::string> kShellHistoryFiles = {
-    ".bash_history", ".zsh_history", ".zhistory", ".history", ".sh_history",
+    ".bash_history",
+    ".zsh_history",
+    ".zhistory",
+    ".history",
+    ".sh_history",
+};
+
+struct HistoryState {
+  std::string content;
+  std::regex bash_timestamp_rx{"^#([0-9]+)$"};
+  std::regex zsh_timestamp_rx{"^: {0,10}([0-9]{1,11}):[0-9]+;(.*)$"};
+  std::string prev_bash_timestamp;
 };
 
-void genShellHistoryFromFile(const std::string& uid,
-                             const boost::filesystem::path& history_file,
-                             QueryData& results) {
-  std::string history_content;
-  if (forensicReadFile(history_file, history_content).ok()) {
```

Comment:
Is moving away from forensic read an issue?

Reply:
Nope, `forensicReadFile` is sort of a misnomer, there's no way to reliably forensically read a file from an online filesystem.